### PR TITLE
Add dateParser, dateFormatter to use functions and move dataTypeMatcher to COLUMN_MAYBE_FUNCTIONS

### PR DESCRIPTION
**Pull request by sdidier-dev**
Changelog:
- Add `dateParser` and `dateFormatter` in COLUMN_DANGEROUS_FUNCTIONS to use functions
- Move `dataTypeMatcher` from COLUMN_MAYBE_FUNCTIONS_NO_PARAMS to COLUMN_MAYBE_FUNCTIONS to use `params`
- Modify ag-grid version to 31.0.2

### CODE REVIEW

**Comment by sdidier-dev** [comment]
@BSd3v for the tests I tried to use the `datepicker` to edit the date, but when the `datepicker` is displayed I can't "inspect" it to find selectors, I can't right-click on it... 🤔
So I sent the text '01172024' (MM-DD-YYYY), but no sure that works all the time as it uses the `datepicker` of the browser, so that depends on locales...

**Comment by sdidier-dev** [comment]
Moreover, do you know why we can't use `.match()` function inline? I wanted to use :
```python
"valueParser": {"function": "params.newValue != null && params.newValue.match('\\d{2}/\\d{2}/\\d{4}') ? params.newValue : null"}
```
but it seems it doesn't like `params.newValue.match('\\d{2}/\\d{2}/\\d{4}')`

Do we need to import something in `src\lib\renderers\customFunctions.js` ?

It seems `.match()` is a String method, but so are `.split()` and `.toUpperCase()` that I found in the docs examples and work:
```python
"valueGetter": {"function": "params.data['animalsString'].split('|')"},
"valueFormatter": {"function": "(params.value).toUpperCase();"}
```

**Comment by BSd3v** [comment]
@sdidier-dev,

As far as the match, to be an inline function, it needs to be simple. I do agree that `.match` seems like it should work. But for some reason it just doesnt pass over. And no, you dont need to import, that is automatically done by the grid.

**Comment by alexcjohnson** [comment]
`.match` accepts a RegExp object, not a string, so in regular js it would be:
```
params.newValue != null && params.newValue.match(/\d{2}\/\d{2}\/\d{4}/) ? params.newValue : null
```
or
```
params.newValue != null && params.newValue.match(new RegExp("\\d{2}\/\\d{2}\/\\d{4}")) ? params.newValue : null
```
But IDK if either of those is compatible with esprima/static-eval.

File: tests/examples/assets/dashAgGridFunctions.js [hunk 1]
```diff
@@ -55,4 +55,85 @@ dagfuncs.startWith = ([filterValues], cellValue) => {
 
 dagfuncs.quickFilterMatcher = (quickFilterParts, rowQuickFilterAggregateText) => {
     return quickFilterParts.every(part => rowQuickFilterAggregateText.match(part));
-}
\ No newline at end of file
+}
+
+dagfuncs.valueParser = (params) => {
+    return params.newValue != null && params.newValue.match('\\d{2}/\\d{2}/\\d{4}')
```
Comment (alexcjohnson) [inline tests/examples/assets/dashAgGridFunctions.js]:
```suggestion
    return params.newValue != null && params.newValue.match(/\d{2}\/\d{2}\/\d{4}/)
```
Here you're in plain JS so the usual JS regex format should work fine.

**Comment by sdidier-dev** [comment]
Thanks Alex!
Actually that comes from AG Grid docs, I also prefer the real regex syntax instead of a string.
I think the issue was the missing `!!`, like `!!params.newValue.match(/\d{2}\/\d{2}\/\d{4}/)` as `.match()` doesn't return a Boolean.
Anyway that's working now 🎉 
I modified the examples and the tests!

- The first grid defines `dataTypeDefinitions` fully in JS
- The second defines it inline with its nested functions:
   - `valueParser`, `valueFormatter `and `dataTypeMatcher `inline
   - `dateParser `and `dateFormatter` in JS

I'm also going to modify the docs 😃 
EDIT: Actually, it's ok for the docs, the existing example uses the full JS and works fine 👌

**Comment by alexcjohnson** [comment]
> Actually that comes from AG Grid docs, I also prefer the real regex syntax instead of a string.

Ah, but look at that, `match` _can_ take a string! I must have screwed up the escaping when I was testing that yesterday because it didn't seem to work but today it does 😅

**Pull request merged**